functionalDB/chapter.md

```diff
@@ -0,0 +1,1097 @@
+
+* Viewed from the outside as a single (i.e., atomic) operation, meaning either all operations succeed or all operations fails
+* Keep the database in a valid state at any given time 
+* Keep the update operations isolated from the outside at any point in time until they are completed. 
+
+The way to answer these requirements is by using a mechanism that takes as an input input the database and a set of operations to be performed, and produces as an output a database whose state reflects the given changes. More than that, all the changes should be seen as an addition of a single layer. 
+
+Performing a transactional operation is done in Clojure using the *Atom* element (this is why we wrap a database with such an element). This element allow us to invoke a single function on it, and all changes to it happen in an atomic and isolated manner. 
+
+As for the function that is to be executed, this function needs to perform all the operations that the user requested, while eventually adding one layer to the database.  To do so we would need to overcome the fact that each of the lifecycle functions adds a layer reflecting the change it inflicts. 
+
+Here we resort to the insight that changes on different attributes (that can be part of either the same or different entities) accumulate when stacking layers, thus a top layer holds all the changes that built the layers below it. Therefore, the solution is to execute each of the user’s operations one after another, each creating a new layer. While doing so, we do not update the databases’s timestamp field. When the last layer is completed, we take only that top layer and place it on the initial database  (leaving all the intermediate layers to pine for the fjords) and only then update the database’s timestamp. All this is happening in the *transact-on-db* function
+
+````clojure
+(defn transact-on-db [initial-db txs]
+     (loop [[tx & rst-tx] txs transacted initial-db]
+       (if tx
+           (recur rst-tx (apply (first tx) transacted (rest tx)))
+           (let [initial-layer  (:layers initial-db)
+                 new-layer (last (:layers transacted))]
+             (assoc initial-db 
+                         :layers (conj  initial-layer new-layer) 
+                         :curr-time (next-ts initial-db) 
+                         :top-id (:top-id transacted))))))
+```` 
+This function is performed on the Atom that wraps the database (to modify the database state as seen by the user), but it can be used also to answer a *what-if* question - as it does not modify the given database, but rather produces a new database with the updated state. These two scenarios are supported by two execution paths that end up in the *transact-on-db* function:

```

Comment:
I have a suggestion for this section. Instead of discussing `what-if` and lazy argument evaluation together, I suggest:

a) Identifying that users can't really call `transact-on-db` directly, because if they provide e.g. a literal list of them in a REPL, the list will be evaluated 

b) We thus provide a higher-level `transact` call, which is actually a macro. The reason why this helps is because we basically get a form of lazy evaluation out of it. Instead of parameterizing `_transact` at this stage, I would explicitly call `*swap*` from within `_transact` to support only this use case.

c) Go through your macro expansion examples for `_transact` to describe how this works in practice.

d) Split out the `what-if` implementation into a short subsection. Here's where you can refactor `_transact` to take a post-update function.

I would also make it more explicit that `transact-on-db` requires an `*Atom*` and not a bare database (why can't `transact-on-db` call `*Atom*` on `initial-db` itself as a first step?)

I'm going to hold off on mutating this section until you've had a chance to respond to this :)